### PR TITLE
rename all node crates

### LINKED ISSUE #2109: Rename all Polkadot node crates
The names are really long and inconsistent.

Every crate in `node/` should be categorized as Core, Network, or Utility. `polkadot-node-core` will become `pnc`, `polkadot-node-network` will become `pnn`, and we'll have `pnu` for utility.

Examples:
`pnc-candidate-backing`
`pnc-candidate-validation`
`pnn-availability-distribution`
`pnn-approval-distribution`
`pnu-subsystem`
`pnu-jobs`
`pnu-subsystem-test`
`pnu-overseer`

### PR DESCRIPTION
**Pull request by coriolinus**
Closes #2109.

This looks huge, but it's all really mechanical stuff: automatic renames and splitting out jobs management and validator discovery from the util crate into their own crates. Should be fairly easy to review.

### CODE REVIEW

**Comment by bkchr** [comment]
What is pnu? And why are there still crates that are prefixed with polakdot?

This renaming is totally random.

**Comment by bkchr** [comment]
Ahh polakdot-node-util.

The naming is still weird. We will still have crates with polakdot prefixes in this repo. Either everything or nothing, no half backed solution in my opinion

File: node/validator-discovery/src/lib.rs [hunk 1]
```diff
@@ -20,29 +20,57 @@ use std::collections::HashMap;
...
 	messages::{AllMessages, NetworkBridgeMessage},
 	SubsystemContext,
 };
 use polkadot_primitives::v1::{Hash, ValidatorId, AuthorityDiscoveryId, SessionIndex};
 use sc_network::PeerId;
-use crate::Error;
+
+/// Errors which can occur when connecting to validators
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
```
Comment (ordian) [inline node/validator-discovery/src/lib.rs:40]: why not use `pnu_subsystem_util::Error` instead of duplicating each variant? only a few variants are actually reachable

alternatively, we can change return types of `pnu_subsystem_util::request_*` to `RuntimeApiError` instead of `util::Error` and reduce the scope of errors

**Comment by coriolinus** [comment]
> Either everything or nothing, no half backed solution in my opinion

Sure, makes sense to me. The originating issue had specified only renaming `node` crates, so that's what's in so far, but I have no objection to extending the convention through the repo.

File: node/validator-discovery/src/lib.rs (hunk 1 above)
Comment (coriolinus) [reply node/validator-discovery/src/lib.rs:40]: I expect rustc to complain if a variant isn't used, and it isn't. That said, I'll look into reducing the number of variants.

**Comment by coriolinus** [comment]
Here is the set of crate names currently in:

```sh
$ fd Cargo.toml -x toml2json | jq -r '.package.name' | sort 
Thu 07 Jan 2021 09:56:35 AM CET
erasure_coding_fuzzer
kusama-runtime
pnc-av-store
pnc-backing
pnc-bitfield-signing
pnc-candidate-selection
pnc-candidate-validation
pnc-chain-api
pnc-proposer
pnc-provisioner
pnc-runtime-api
pnn-availability-distribution
pnn-bitfield-distribution
pnn-bridge
pnn-collator-protocol
pnn-pov-distribution
pnn-protocol
pnn-statement-distribution
pnu-collation-generation
pnu-jaeger
pnu-jobs
pnu-overseer
pnu-primitives
pnu-service
pnu-subsystem
pnu-subsystem-test-helpers
pnu-subsystem-util
pnu-test-client
pnu-test-service
pnu-validator-discovery
polkadot
polkadot-cli
polkadot-core-primitives
polkadot-erasure-coding
polkadot-parachain
polkadot-primitives
polkadot-rpc
polkadot-runtime
polkadot-runtime-common
polkadot-runtime-parachains
polkadot-statement-table
polkadot-test-runtime
polkadot-validation
rococo-runtime
test-parachain-adder
test-parachain-adder-collator
test-parachain-halt
test-parachains
westend-runtime
xcm
xcm-builder
xcm-executor
```

Here are the renaming conventions I'd use:

- `*-runtime` -> `runtime-*`
- `polkadot-*` -> `pdot-*`

Everything else I'd leave the same. @bkchr do you have any other rename preferences, or would applying those two rename rules satisfy you?

File: node/validator-discovery/src/lib.rs (hunk 1 above)
Comment (coriolinus) [reply node/validator-discovery/src/lib.rs:40]: I think add9c59e resolves this.

**Comment by rphmeier** [comment]
@bkchr The issue was just about renaming and making consistent names for crates in `polkadot/node`. e.g. `polkadot-node-core-approval-voting` is way too long. I don't think that having some crates with the name `polkadot-X` is actually a problem that we need to solve, as the names are short already.

**Comment by bkchr** [comment]
> @bkchr The issue was just about renaming and making consistent names for crates in `polkadot/node`. e.g. `polkadot-node-core-approval-voting` is way too long. I don't think that having some crates with the name `polkadot-X` is actually a problem that we need to solve, as the names are short already.

Where is the problem that the name is long? I personally find some consistent naming scheme much more important than the length of the crate names. Look at substrate, there exists a consistent naming scheme since some time and it also made names shorter.

Tldr, I'm fine with renaming, but than we should make it consistent over all crates.

**Comment by rphmeier** [comment]
Do you have any suggestion for how we should rename the other `polkadot` crates?

**Comment by coriolinus** [comment]
@bkchr @rphmeier Is this PR worth updating/merging, or should we drop it? I don't know what further work is desired at this point.

**Comment by rphmeier** [comment]
Probably just needs some critical thought and then we can move forward with this. We need to correct the inconsistencies between crate names and ideally make them shorter.

**Comment by rphmeier** [comment]
I think we could just rename everything to `pn` now. Still not sure what to do about all the non-node crates. I'm fine with leaving them all. I think we should put every crate into either `node`, `runtime`, or `util`. TBH the only things that might go in `util` are `erasure-coding`, `primitives`, and `xcm`.

**Comment by coriolinus** [comment]
The current set of crate names:

```sh
$ fd Cargo.toml -x toml2json | jq -r '.package.name' | sort 
Fri 29 Jan 2021 10:13:43 AM CET
erasure_coding_fuzzer
pdot-cli
pdot-core-primitives
pdot-erasure-coding
pdot-parachain
pdot-primitives
pdot-rpc
pdot-runtime-common
pdot-runtime-parachains
pdot-statement-table
pdot-test-runtime
pdot-validation
pnc-av-store
pnc-backing
pnc-bitfield-signing
pnc-candidate-selection
pnc-candidate-validation
pnc-chain-api
pnc-proposer
pnc-provisioner
pnc-runtime-api
pnn-availability-distribution
pnn-bitfield-distribution
pnn-bridge
pnn-collator-protocol
pnn-pov-distribution
pnn-protocol
pnn-statement-distribution
pnu-collation-generation
pnu-jaeger
pnu-jobs
pnu-overseer
pnu-primitives
pnu-service
pnu-subsystem
pnu-subsystem-test-helpers
pnu-subsystem-util
pnu-test-client
pnu-test-service
pnu-validator-discovery
polkadot
runtime-kusama
runtime-polkadot
runtime-rococo
runtime-westend
test-parachain-adder
test-parachain-adder-collator
test-parachain-halt
test-parachains
xcm
xcm-builder
xcm-executor
```

@rphmeier I believe that this addresses the goal of making the names shorter. As for the general restructuring, this is how I interpret your instructions.

- `polkadot` will still live in the directory `src`
- `pdot-erasure-coding` will be renamed to `pnu-erasure-coding` and move into an appropriate directory
- `pdot-(*)` will live in the directory `polkadot/\1`
- `pnc-(*)` will live in the directory `node/core/\1`
- `pnn-(*)` will live in the directory `node/network/\1`
- `pnu-(*)` will live in the directory `node/util/\1`
- `runtime-(*)` will live in the directory `runtime/\1`
- `test-(*)` will live in the directory `test/\1`
- `xcm` will be renamed to `pnu-xcm` and move into an appropriate directory
- `xcm-*` will be renamed to `pnu-xcm-*` and move into an appropriate directory

This gives us the following top-level directories:

- src
- polkadot
- node
- runtime
- test

File: runtime/polkadot/Cargo.toml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [package]
-name = "polkadot-runtime"
+name = "runtime-polkadot"
```
Comment (bkchr) [inline runtime/polkadot/Cargo.toml:2]: I would keep the `polkadot-runtime`, `kusama-runtime` etc names

**Review by bkchr — COMMENTED** [review]
There are everywhere crate renamings leftovers. These should be fixed as well.

`xcm` isn't a node util. This is a pure runtime crate.

**Comment by bkchr** [comment]
And stuff like `pdot-cli`. Why isn't that `pn-cli` as this is client/node code?

I think we also should not name anything `pdot`. This is again some random prefix you don't know what falls into this category.

In substrate it is really simple `sc` == `substrate-client`, so anything that is on the node side. `sp` == `substrate-primitives`, so anything that is used in the runtime/node.

File: runtime/polkadot/Cargo.toml (hunk 2 above)
Comment (coriolinus) [reply runtime/polkadot/Cargo.toml:2]: The reason for renaming them is to provide some straightforward mapping between the location in the directory tree and the package name; improving that mapping is AFAICT one of the major motivations for this issue.

**Comment by coriolinus** [comment]
@bkchr Not sure what you mean by leftovers remaining. I'm still working through the list of rename rules I proposed a few hours ago, so if you mean that those are not all done yet, I'm aware of that and still working on it.

It's straightforward to rename `xcm` as `runtime-xcm` and move it into the `runtime` folder. Good point there. I was just going from @rphmeier 's suggestion.

> I think we also should not name anything `pdot`. 

I'm happy to apply any naming rule you like, there, including just leaving them all named `polkadot-X`. I just don't have a better one available right now than bulk-renaming all `polkadot-x` to something else, and I don't know enough to apply a finer-grained rule to only some of them.

**Comment by bkchr** [comment]
> I'm happy to apply any naming rule you like, there, including just leaving them all named `polkadot-X`. I just don't have a better one available right now than bulk-renaming all `polkadot-x` to something else, and I don't know enough to apply a finer-grained rule to only some of them.

Which one do you don't know?

**Comment by coriolinus** [comment]
For example:

```sh
$ fd Cargo.toml -x toml2json | jq -r '.package.name' | rg 'primitive' | sort
Fri Jan 29 15:41:46 CET 2021
pdot-core-primitives
pdot-primitives
pnu-primitives
```

You suggested `pn-cli`. That implies a rename rule which would create `pn-primitives` and `pn-core-primitives`. However, in both those cases, there are `#[cfg(feature="std")` gates all over the place, implying that those aren't just node modules. The general result is that we have (for now) these `pdot` modules:

```
pdot-cli
pdot-core-primitives
pdot-erasure-coding
pdot-parachain
pdot-primitives
pdot-rpc
pdot-runtime-common
pdot-runtime-parachains
pdot-statement-table
pdot-test-runtime
pdot-validation
```

which should map onto some set of abbreviations. I don't know what the set is. `pdot` was an attempt to come up with a universal abbreviation. You don't want it, which is fine, but you haven't suggested yet either a universal abbreviation I can use for them or even a set of abbreviations to which I can assign them.

Even given a set of abbreviations to use, some of these are just not obvious. Say we have `pn` for polkadot-node and `pr` for polkadot-runtime. Which applies to `pdot-rpc`? Best guess is that it's a node thing, but I have very little confidence in that.

**Comment by bkchr** [comment]
> pdot-cli

This is clearly node.

> pdot-core-primitives
> pdot-parachain
> pdot-primitives

All these 3 should be refactored probably into some common crate. There also exists the `pnc-primitives`. Primitives over primitives over primitives.

> pdot-erasure-coding

pnu

> pdot-rpc

Also clearly node.

> pdot-runtime-common
> pdot-runtime-parachains

What is with `pr-parachains` and `pr-common` for polkadot-runtime.

> pdot-statement-table

That is probably node-util

> pdot-test-runtime

You call the other runtimes also just `name-runtime`, same applies here than.

> pdot-validation

Node util
